### ISSUE TITLE
[Alarms & Timers] Option for "Save on back"

### ISSUE THREAD

**Issue by alessandrococco** [alessandrococco/BangleApps#5]
> When I change the config (mostly en-/disable) of an alarm or timer, I don't want to navigate to the Save >-button. I would prefer to just return with the back-icon in the top left corner and have the changes saved.
If enabled the "Save on Back"-option should switch the behavior of the menu entries Back > <-> Save > and rename the Save > to Cancel >.

https://github.com/espruino/BangleApps/pull/1707#issuecomment-1102764309

**Comment by storm64** [comment]
Maybe it's more suitable to discuss the new feature here. ;)

> @storm64 I just implemented the "Save on Back", you can find it in [this branch](https://github.com/alessandrococco/BangleApps/tree/alarms-save-on-back) - Feel free to do some tests, a feedback is appreciated!

Couldn't start the alarm app for now due to missing locale. I think you need to add your changes in the locale library/module in the [espruino/Espruino: libs/js/banglejs/locale.js](https://github.com/espruino/Espruino/blob/master/libs/js/banglejs/locale.js), too.

```
Uncaught Error: Function "dowAll" not found!
 at line 9 col 36 in alarm.app.js
require("locale").dowAll(2).map((day,index)=>dow&(1<<index)?da...
                    ^
in function "decodeDOW" called from line 8 col 552 in alarm.app.js
...xt+"\n"+decodeDOW(alarm.dow);}menu[txt]={value:"\0"+atob(ala...
                              ^
in function called from line 8 col 792 in alarm.app.js
...;else editAlarm(idx,alarm);}};});if(WIDGETS["alarm"])WIDGETS["ala...
                                   ^
in function "showMainMenu" called from line 15 col 14 in alarm.app.js
showMainMenu();
             ^
```


> App loader link: https://alessandrococco.github.io/BangleApps/?id=alarm (v0.24)

Overlooked this. :sweat_smile:

**Comment by alessandrococco** [comment]
Hi @storm64,
you should update the Languages app to v0.18 ;-)

**Comment by storm64** [comment]
> Hi @storm64,
you should update the Languages app to v0.18 ;-)

Until now I haven't used the locale app, therefore the firmware internal library has been used.

Installed locale and the "Save on Back"-feature works charming, thanks. :+1:

---

I like your idea to display the dow on the alarm/timer selection screen, but for now I'm not totally sure if I like it that much. (Maybe I need to get used to it. ;)
Would suggest to hide the dow if all days are selected and like to have a en-/disable option in the settings, too.

There are two other ideas I have in mind to make the dow visible:
 1. Only display the count of selected days (if lower 7):
    - `S_____S` -> `Alarm 2d `:arrows_counterclockwise:
    - `_MTWTF_` -> `Alarm 5d `:arrows_counterclockwise:
    - `SMTWTFS` -> `Alarm `:arrows_counterclockwise:
 2. Show special identifier for weekend and weekdays:
    - `S_____S` -> `Alarm WE `:arrows_counterclockwise:
    - `_MTWTF_` -> `Alarm WD `:arrows_counterclockwise:

The second version could be combined with your idea by displaying the distinct days except for WE, WD and if all days are selected.

**Comment by alessandrococco** [comment]
> Installed locale and the "Save on Back"-feature works charming, thanks. 👍

Great!

> Would suggest to hide the dow if all days are selected and like to have a en-/disable option in the settings, too.
> There are two other ideas I have in mind to make the dow visible:
> 1. Only display the count of selected days (if lower 7)
> 2. Show special identifier for weekend and weekdays

I could add some options. Maybe a new menu `Days of Week` with:

- Show Days (eg.: `Alarm 9:00\n_MTWTF_`)
- Show Days Count (`Alarm 9:00 5d`)
- Hide

The problem is the limited space available: the max length to keep the text on one line is 12. `Alarm 12:00 5d` is 14 (16 in Italian). Moreover there is the 🔁 icon :-/


Thank you for your feedback!

**Comment by storm64** [comment]
> I could add some options. Maybe a new menu Days of Week

This would be great.

> The problem is the limited space available: the max length to keep the text on one line is 12. Alarm 12:00 5d is 14 (16 in Italian). Moreover there is the 🔁 icon :-/

:thinking: This is a good point...
This in mind, I don't think the count option is that good and might only be helpful on shortening the Alarm to just A: `A 5d 12:00`... I don't know.

Another idea would be to replace the Alarm name with the dow: `_MTWTF_ 9:00` this might fit most of the time?

And I would like to only have the dow displayed if at least one day is deselected.

---

I really like your work an taking a part in it.

**Comment by alessandrococco** [comment]
Thanks! 

> Another idea would be to replace the Alarm name with the dow: _MTWTF_ 9:00 this might fit most of the time?

This seems to be the best idea. This should also help to handle duplicated alarms.

> And I would like to only have the dow displayed if at least one day is deselected.

Not sure about this, I think it's confusing. Eg.:

`_MTWTF_ 9:00`  
`Alarm 9:00` 
 
Does the `Alarm 9:00` run every day? Or only once?

---

The default "repeat" is also confusing: the doc says `rp: true, // repeat the alarm every day?`; Will an alarm configured as `_MT____ 9:00 🔁`  fire on Wednesday? (Maybe if the user set the repeat to Yes the app must force SMTWTFS 🤔)

**Comment by storm64** [comment]
> Does the Alarm 9:00 run every day? Or only once?

I would say this is depending on the repeat setting:

If a Alarm is not repeating it will only trigger once if activated. If repeating is enabled it will reenable itself after getting triggered.

Therefore it should be distinct if now dow is displayed depending on the repeating icon and, of cause, the status:
1. `_MTWTF_ 9:00`
    - rp = true - > The alarm will be triggered on each selected day.
    - rp = false - > The alarm will be triggered and get disabled on the next selected day.
2. `Alarm 9:00`
    - rp = true - > The alarm will be triggered every day.
    - rp = false - > The alarm will be triggered and get disabled on the next day.
3. `_MT____ 9:00`
    - rp = true - > The alarm will be triggered on each Mon- and Tuesday.
    - rp = false - > The alarm will be triggered and disabled on the next Mon- or Tuesday.

**Comment by alessandrococco** [comment]
Hi @storm64, thank you again.

I sent a PR for the `Save on Back` *without* the dow stuff (https://github.com/espruino/BangleApps/pull/1748)

I'll work on the dow in the next few days!

**Comment by alessandrococco** [comment]
New PR https://github.com/espruino/BangleApps/pull/1773
